compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/FirLocalVariableAssignmentAnalyzer.kt

```diff
@@ -0,0 +1,349 @@
+                    forEach { it.children.addIfNotNull(result) }
+                    return result
+                }
+            }
+        }
+
+
+        private class MiniCfgBuilder : FirVisitor<Unit, MiniCfgBuilder.MiniCfgData>() {
+            override fun visitElement(element: FirElement, data: MiniCfgData) {
+                element.acceptChildren(this, data)
+            }
+
+            override fun visitAnonymousFunction(anonymousFunction: FirAnonymousFunction, data: MiniCfgData) {
+                handleFunctionFork(anonymousFunction, data)
+            }
+
+            override fun visitSimpleFunction(simpleFunction: FirSimpleFunction, data: MiniCfgData) {
+                handleFunctionFork(simpleFunction, data)
+            }
+
+            private fun handleFunctionFork(function: FirFunction<*>, data: MiniCfgData) {
+                val currentFlow = data.flow ?: return
+                val functionFork = currentFlow.fork()
+                data.flow = functionFork
+                function.acceptChildren(this, data)
+
+                // Only retain local variables declared above the current scope. This way, any local variables declared inside the
+                // function will effectively be treated as distinct variables and, hence, stable (Of course, for nested lambda, things would
+                // just work because inside the lambda assigned local variables are tracked by different nodes).
+                functionFork.assignedLocalVariables.retainAll(data.variableDeclarations.flatMap { it.values })
+                // Create another fork for the normal execution
+                val normalExecution = currentFlow.fork()
+                data.localFunctionToAssignedLocalVariables[function] =
+                    AssignedLocalVariables(normalExecution.assignedLocalVariables, functionFork.assignedLocalVariables)
+                data.flow = normalExecution
+            }
+
+            override fun visitWhenExpression(whenExpression: FirWhenExpression, data: MiniCfgData) {
+                val flow = data.flow ?: return
+                val visitor = this
+                with(whenExpression) {
+                    calleeReference.accept(visitor, data)
+                    val subjectVariable = this.subjectVariable
+                    if (subjectVariable != null) {
+                        subjectVariable.accept(visitor, data)
+                    } else {
+                        subject?.accept(visitor, data)
+                    }
+                    val childFlows = branches.mapNotNull {
+                        data.flow = flow.fork()
+                        it.accept(visitor, data)
+                        data.flow
+                    }
+                    data.flow = (childFlows + flow).toSet().join()
+                }
+            }
+
+            override fun visitReturnExpression(returnExpression: FirReturnExpression, data: MiniCfgData) {
+                data.flow = null
```

Comment:
It's not for unreachable code. It's only to mark the execution has completed and won't be joined by other execution. For example, with the following
```
if(blah) {
  return
}
```
This is to prevent the returned flow from being joined after the if condition.

Fork due to lambda doesn't need special handling at all because the exeuction path never joins at the end. So there is no need to handle implicit return in lambda: a lambda is always "run-away"

For throw, things are much more complicated due to the possibility of catching them (where is the catch? does each catch throw at the end? etc). Hence, I am not doing anything throw. The downside is that it would reject more valid code. For example.

```
var x: String? = null
x = ""
if (blah) {
  runLambda {
    x.length // report as unstable
  }
  throw Exception()
}
x = null
```

It should have been stable but it's not. Replacing `throw` with `return` makes it stable. I didn't go down the rabbit hole to handle all of these. But maybe we can add a TODO here?

Reply:
Leaving some TODOs to not forget about that sounds good, and actually good enough for now.